### ISSUE TITLE
`rustc_macros`: Make it possible to derive both `Diagnostic` and `LintDiagnostic` on the same type

### ISSUE THREAD

**Issue by fmease** [rust-lang/rust#125169]
In #117164 I had to copy the diagnostic structs `TyParamFirstLocal` and `TyParamSome` and define the separate structs `TyParamFirstLocalLint` and `TyParamSomeLint` to be able to use the translatable diagnostics `hir_analysis_ty_param_first_local` and `hir_analysis_ty_param_some` as both a `Diagnostic` and a `LintDiagnostic`:

https://github.com/rust-lang/rust/blob/b21b74b5e673b9102502e8f69f378313d901746c/compiler/rustc_hir_analysis/src/errors.rs#L1360-L1406

In #116829, had to copy the diagnostic struct `ReprConflicting` and define the separate struct `ReprConflicting` to be able to use the translatable diagnostic `passes_repr_conflicting` as both a `Diagnostic` and a `LintDiagnostic`:

https://github.com/fmease/rust/blob/d0b99e3efe62b5acc107da9c84331eae6cbe5a0d/compiler/rustc_passes/src/errors.rs#L561-L570

In all three cases, I could've probably turned the derivation of `LintDiagnostic` into a manual impl but that doesn't seem very enticing either for various reasons.

For context, this situation arises whenever we want to emit a diagnostic for something but can't report a hard error unconditionally in all cases for backward compatibility and therefore have to emit the very same diagnostic at different “levels of severity” depending on a set of condition. With *level of severity* I'm specifically referring to the set {hard error, lint} here (where *lint* has its own level of course).

---

More concretely, the reason why you can't just `#[derive(Diagnostic, LintDiagnostic)] struct Ty/*...*/` is because of `#[primary_span]`: If the diagnostic struct contains a `#[primary_span]` (which it does most of the time) for the derive macro `Diagnostic`, the derive macro `LintDiagnostic` rejects `#[primary_span]` rendering the two derives incompatible with each other.

Individually, `LintDiagnostic` rejecting `#[primary_span]` makes sense because the span(s) for a lint are to be provided to the function that emits the lint like `TyCtxt::emit_spanned_lint`.

---

There are probably multiple ways to approach this but I'm not super familiar with internals of rustc's linting API. Anyways, I'd like to see this “just work” because this won't be the last time this case will arise.

**Comment by fmease** [comment]
I'll probably investigate getting rid of `emit_spanned_lint` in favor of `emit_lint` (no  span / multi-span parameter) and storing the span(s) of `LintDiagnostic`s in the lint diagnostic structs themselves via `#[primary_span]` which seems to be the most logical approach. However, I'll require updating all lint diagnostic structs.

cc @nnethercote

**Comment by compiler-errors** [comment]
Yep, I think that `LintDiagnostics` should be responsible for providing the `primary_span`. This should be pretty easy to do 👍

**Comment by fmease** [comment]
Ah, that's annoying. We can't simply leverage `Diag::span` (inside `LintDiagnostic::decorate_lint`) unlike `Diagnostic` because `lint_level` (`lint_level_impl`) needs access to the (multi) span before `decorate`'ing a `Diag` to suppress lints whose spans come from an external macro (and to disable suggestions):

https://github.com/rust-lang/rust/blob/4eedf7385bcfbaa83f91bd86a282de2caeed9b07/compiler/rustc_middle/src/lint.rs#L333-L351

By the way, I've already migrated all lint diagnostic structs to the new system.

The only solution I see is adding a `span` method to trait `LintDiagnostic` (similar to the existing `msg` method) which really wanted to avoid because it makes the API more complicated and bloats the generated code.

**Comment by fmease** [comment]
Right now, my patched rustc fails to build std because of that >.<

**Comment by fmease** [comment]
https://github.com/rust-lang/rust/pull/125208#issuecomment-2437651424

> Yesterday I coincidentally stumbled upon [this Zulip post](https://rust-lang.zulipchat.com/#narrow/channel/336883-i18n/topic/.23100717.20diag.20translation/near/472701978) from 2024-09-25 which essentially declares the current diagnostic translation effort (#100717) unfruitful and unmaintained. While I only read this post yesterday, I have been aware of this fact for some time.
> 
> Therefore I don't want to burden davidtwco with re-reviewing >1000 lines and I'm also fine with closing this PR.
> 
> Of course, not a lot has changed. It's essentially a rebase from a 5 month old master. I stalled the rebase until now because I knew that this approach (of making lint diagnostics responsible or providing their primary span) does not scale to early (buffered) lints. I came to realize that when reviewing PR #124417 (which made early lints translatable) and implementing PR #125410 (which won back some of the perf regressions caused by the former).
> 
> You say it doesn't scale to early (buffered) lints, so how did you succeed in rebasing then while keeping perf in check? Well, I've cheated a little bit. The statement "Lints diagnostics are responsible for providing their primary span" now only only holds for non-early lints. The early lint diagnostics don't provide a `#[primary_span]` / return `None` for `LintDiagnostic::span(&self) -> Option<MultiSpan>` with the **hidden assumption/knowledge**[1](#user-content-fn-1-a363f01f216926798ebcd37bcbd0b2d1) that it's supplied separately in `opt_span_lint_with_diagnostics` from `BufferedEarlyLint.span`.
> 
> For perf reasons, we can't eagerly "lower" `BuiltinLintDiag` to lint diagnostics ("`impl LintDiagnostic`") inside `opt_span_lint_with_diagnostics` via `decorate_lint`[2](#user-content-fn-2-a363f01f216926798ebcd37bcbd0b2d1). However, only if we lower eagerly, we can truly make early lint diagnostics responsible for providing their primary span – and that's off the table. Re. perf, the lowering step is expensive and therefore we delay this step as far back as possible. Namely we don't want to lower lint diagnostics that end up not getting emitted (due to lint levels). We achieve this by letting `lint_level` decide … which takes a `Span` that is used in its decision process! We need the primary span(s) before lowering to lint diagnostics, therefore lint diagnostics cannot possibly provide their primary span!
> 
> This creates a divide between early and late `LintDiagnostic`s which is unacceptable in my opinion. I wanted bring `{,Lint}Diagnostic` closer together for ease of understanding and use. This simply doesn't work if `LintDiagnostic` itself is divided. **As such, I no longer believe in this PR**.
> ## Footnotes
> 
> 1. Well, I did add comments in the `LintDiagnostic::span` cases but not in the `#[derive(LintDiagnostic)]` ones (there are simply too many). [↩](#user-content-fnref-1-a363f01f216926798ebcd37bcbd0b2d1)
> 2. See the merged commit https://github.com/rust-lang/rust/pull/125410/commits/37bf2d2dabdbdce9473b0fed1708fcbf31ba9c1a which reverted this. [↩](#user-content-fnref-2-a363f01f216926798ebcd37bcbd0b2d1)
